File: 프로그래머스/알고리즘 고득점 Kit/깊이,너비 우선 탐색(DFS,BFS)/Level 2/게임 맵 최단거리/게임 맵 최단거리.py
def solution(maps):
    visit = [[0 for i in range(len(maps[0]))] for i in range(len(maps))]

    di=[-1, 0, 1, 0]; dj=[0, 1, 0, -1]
    visit[0][0] = 1
    q = [(0, 0)]

    while len(q):
        tmp_i = q[0][0]
        tmp_j = q[0][1]
        q.pop(0)

        for i in range(4):
            move_i = tmp_i+di[i]
            move_j = tmp_j+dj[i]

            if move_i<0 or move_i>=len(maps): continue
            if move_j<0 or move_j>=len(maps[0]): continue

            if maps[move_i][move_j]==1 and visit[move_i][move_j]==0:
                q.append([move_i, move_j])
                visit[move_i][move_j] = visit[tmp_i][tmp_j]+1

    if visit[-1][-1] > 0: return visit[-1][-1]
    else: return -1

File: 프로그래머스/알고리즘 고득점 Kit/깊이,너비 우선 탐색(DFS,BFS)/Level 2/게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>
using namespace std;

int solution(vector<vector<int>> maps){
    vector<vector<int>> visit;
    for(int i=0; i<maps.size(); i++){
        vector<int> v(maps[0].size(), 0);
        visit.push_back(v);
    }
    
    // 상우하좌
    int di[4]={-1, 0, 1, 0}, dj[4]={0, 1, 0, -1};
    visit[0][0] = 1;
    queue<pair<int, int>> q;
    q.push(make_pair(0, 0));
    
    while(q.size()){
        int tmp_i = q.front().first;
        int tmp_j = q.front().second;
        q.pop();
        
        for(int i=0; i<4; i++){
            int move_i = tmp_i+di[i];
            int move_j = tmp_j+dj[i];
            
            if(move_i<0 || move_i>=maps.size()) continue;
            if(move_j<0 || move_j>=maps[0].size()) continue;
            
            if(maps[move_i][move_j] && visit[move_i][move_j]==0){
                q.push(make_pair(move_i, move_j));
                visit[move_i][move_j] = visit[tmp_i][tmp_j]+1;
            }
        }
    }
    
    int answer = visit[maps.size()-1][maps[0].size()-1];
    return answer ? answer : -1;
}
